File: RSA-NoLib.py
import math

# Fonction pour calculer le PGCD 
def pgcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

# Fonction de l'algorithme d'Euclide étendu (pour trv l'inverse modulaire)
def bezout(a, b):
    x0, x1, y0, y1 = 1, 0, 0, 1
    while b != 0:
        q, a, b = a // b, b, a % b
        x0, x1 = x1, x0 - q * x1
        y0, y1 = y1, y0 - q * y1
    return a, x0, y0  # a = PGCD, x0 = coefficient de Bézout

# Fonction d'exponentiation modulaire (pr chiffrer/déchiffrer)
def puissance(base, exponent, modulo):
    result = 1
    base = base % modulo
    while exponent > 0:
        if (exponent % 2) == 1:  # Si exponent est impair
            result = (result * base) % modulo
        exponent = exponent >> 1  # Divise exponent par 2
        base = (base * base) % modulo
    return result

# Conversion du texte en nombre
def to_number(text):
    number = 0
    for char in text:
        number = number * 256 + ord(char)  # Chaque caractère est codé sur 1 octet
    return number

# Conversion du nombre en texte
def from_number(number):
    chars = []
    while number > 0:
        chars.append(chr(number % 256))  # On récupère les caractères un par un
        number //= 256
    return ''.join(reversed(chars))

# Système RSA

# Deux grands nombres premiers p et q
p = 12371247788555446551320478301289
q = 54684665222227225444533546543532501
n = p * q
print(n, "égal à p*q")

# On calcule φ(n) = (p-1)*(q-1)
z = (p - 1) * (q - 1)

# Choisir e, qui doit être premier avec φ(n)
e = 525567875157816551

# Vérification que e est premier avec φ(n
def verification(e, z):
    if pgcd(e, z) != 1:
        print("Erreur : e n'est pas premier avec (p-1)*(q-1)")
    else:
        print('Tout va bien, e est bien premier avec (p-1)*(q-1)')

verification(e, z)

# Calculer d, l'inverse de e mod φ(n)
_, d, _ = bezout(e, z)
d = d % z  # On prend l'inverse modulaire de e

print("La relation de Bézout:")
print(f"e * d ≡ 1 mod φ(n) où d = {d}")
print(d, "est l'inverse de e")

# Le message à chiffrer
Message = 'Donnez moi un poste de pentester svp'
print(Message)

# Convertir le message en nombre
M = to_number(Message)
print(M, "est le message numérisé M")

# Bob chiffre le message
C = puissance(M, e, n)
print(C, "est le message chiffré")

# Alice déchiffre le message
D = puissance(C, d, n)
print(D, "devrait être égal à M")

# Vérification que le déchiffrement est correct
if M == D:
    print("Le déchiffrement a réussi, le message déchiffré est correct.")
else:
    print(f"Le message déchiffré ({D}) est différent du message original ({M}).")

# Retrouver le message à partir du nombre déchiffré
message_retrouve = from_number(D)
if message_retrouve is not None:
    print('Et enfin: ', message_retrouve)
else:
    print("Impossible de retrouver le message original à partir du nombre déchiffré.")

File: RSA-NoLib.c
#include <stdio.h>
#include <stdlib.h>
#include <gmp.h>

// Fonction pour calculer le PGCD
void pgcd(mpz_t result, mpz_t a, mpz_t b) {
    mpz_gcd(result, a, b);
}

// Algorithme d'Euclide étendu pour trouver l'inverse modulaire
// Calcul de l'inverse modulaire (bezout) 
void bezout(mpz_t d, mpz_t a, mpz_t b, mpz_t x, mpz_t y) {
    mpz_t temp, q, r, x0, x1, y0, y1;
    mpz_inits(temp, q, r, x0, x1, y0, y1, NULL);

    mpz_set_ui(x0, 1);
    mpz_set_ui(x1, 0);
    mpz_set_ui(y0, 0);
    mpz_set_ui(y1, 1);

    while (mpz_cmp_ui(b, 0) != 0) {
        gmp_printf("a = %Zd, b = %Zd\n", a, b);  // Debugage, affichage de a et b
        mpz_fdiv_qr(q, r, a, b);

        if (mpz_cmp_ui(r, 0) == 0) {
            gmp_printf("Le reste devient zéro, arrêt de l'algorithme.\n");
            break;
        }

        mpz_set(a, b);
        mpz_set(b, r);

        mpz_mul(temp, q, x1);
        mpz_sub(x, x0, temp);
        mpz_set(x0, x1);
        mpz_set(x1, x);

        mpz_mul(temp, q, y1);
        mpz_sub(y, y0, temp);
        mpz_set(y0, y1);
        mpz_set(y1, y);
    }

    mpz_set(d, x0);
    if (mpz_cmp_ui(d, 0) < 0) {
        mpz_add(d, d, b);
    }

    gmp_printf("Valeur de d après bezout : %Zd\n", d);  // Debugage, Vérification de d
    mpz_clears(temp, q, r, x0, x1, y0, y1, NULL);
}

// Fonction d'exponentiation modulaire
void puissance(mpz_t result, mpz_t base, mpz_t exponent, mpz_t modulo) {
    mpz_t temp;
    mpz_init(temp);
    mpz_set_ui(result, 1);
    mpz_mod(temp, base, modulo);

    while (mpz_cmp_ui(exponent, 0) > 0) {
        if (mpz_tstbit(exponent, 0)) {
            mpz_mul(result, result, temp);
            mpz_mod(result, result, modulo);
        }
        mpz_mul(temp, temp, temp);
        mpz_mod(temp, temp, modulo);
        mpz_fdiv_q_2exp(exponent, exponent, 1);
    }
    mpz_clear(temp);
}

// Conversion du texte en nombre
void to_number(mpz_t result, const char *text) {
    mpz_set_ui(result, 0);
    while (*text) {
        mpz_mul_ui(result, result, 256);
        mpz_add_ui(result, result, (unsigned char)(*text));
        text++;
    }
}

// Conversion du nombre en texte
void from_number(mpz_t number, char *output) {
    size_t i = 0;
    mpz_t temp;
    mpz_init(temp);

    while (mpz_cmp_ui(number, 0) > 0) {
        mpz_fdiv_qr_ui(temp, number, number, 256);
        output[i++] = (char)mpz_get_ui(temp);
    }
    output[i] = '\0';

    for (size_t j = 0; j < i / 2; j++) {
        char tmp = output[j];
        output[j] = output[i - j - 1];
        output[i - j - 1] = tmp;
    }

    mpz_clear(temp);
}

int main() {
    mpz_t p, q, n, z, e, d, M, C, D, gcd;
    mpz_inits(p, q, n, z, e, d, M, C, D, gcd, NULL);

    mpz_set_str(p, "12371247788555446551320478301289", 10);
    mpz_set_str(q, "54684665222227225444533546543532501", 10);
    mpz_mul(n, p, q);
    gmp_printf("n = p * q = %Zd\n", n);

    mpz_t p_original, q_original;
    mpz_inits(p_original, q_original, NULL);
    mpz_set(p_original, p);
    mpz_set(q_original, q);

    mpz_sub_ui(p, p, 1);
    mpz_sub_ui(q, q, 1);
    mpz_mul(z, p, q);
    gmp_printf("φ(n) = %Zd\n", z);

    mpz_set_str(e, "525567875157816551", 10);

    // Vérification du PGCD de e et φ(n)
    pgcd(gcd, e, z);
    gmp_printf("PGCD de e et φ(n) : %Zd\n", gcd);
    if (mpz_cmp_ui(gcd, 1) != 0) {
        printf("Erreur : e n'est pas premier avec φ(n)\n");
        mpz_clears(p_original, q_original, gcd, NULL);
        return 1;
    } else {
        printf("Tout va bien, e est bien premier avec φ(n)\n");
    }

    bezout(d, e, z, p_original, q_original);
    mpz_mod(d, d, z);
    gmp_printf("L'inverse de e mod φ(n) est : d = %Zd\n", d);

    // Vérification si d est positif
    if (mpz_cmp_ui(d, 0) < 0) {
        mpz_add(d, d, z); 
    }

    const char *Message = "Donnez moi un poste de pentester svp";
    gmp_printf("Message original : %s\n", Message);

    to_number(M, Message);
    gmp_printf("Message numérisé : M = %Zd\n", M);

    puissance(C, M, e, n);
    gmp_printf("Message chiffré : C = %Zd\n", C);

    puissance(D, C, d, n);
    gmp_printf("Message déchiffré (numérisé) : D = %Zd\n", D);

    if (mpz_cmp(M, D) == 0) {
        printf("Le déchiffrement a réussi, le message déchiffré est correct.\n");
    } else {
        printf("Erreur : le message déchiffré est incorrect.\n");
    }

    char message_retrouve[256];
    from_number(D, message_retrouve);
    printf("Le message déchiffré est : %s\n", message_retrouve);

    mpz_clears(p, q, n, z, e, d, M, C, D, gcd, p_original, q_original, NULL);

    return 0;
}
